framework/src/media/OutputHandler.cpp

```diff
@@ -0,0 +1,357 @@
+	if (mIsWorkerAlive) {
+		// Setup flag,
+		mIsWorkerAlive = false;
+
+		// Worker may be blocked in buffer reading.
+		mBufferWriter->setEndOfStream();
+
+		// Wake worker up,
+		wakenWorker();
+
+		// Join thread.
+		pthread_join(mWorker, NULL);
+	}
+}
+
+void *OutputHandler::workerMain(void *arg)
+{
+	medvdbg("OutputHandler::workerMain()\n");
+
+	if (arg == nullptr) {
+		meddbg("%s[line : %d] Fail : arg is nullptr\n", __func__, __LINE__);
+		return NULL;
+	}
+
+	auto stream = static_cast<OutputHandler *>(arg);
+	auto worker = stream->mOutputDataSource;
+
+	while (stream->mIsWorkerAlive) {
+		// Waken up by a writing/flushing/stopping operation
+		stream->sleepWorker();
+
+		// Worker may be stoped
+		if (!stream->mIsWorkerAlive) {
+			break;
+		}
+
+		auto size = stream->mBufferReader->sizeOfData();
+		if ((size >= stream->mStreamBuffer->getThreshold()) ||
+			(size > 0 && stream->mIsFlushing)) {
+			auto buf = new unsigned char[size];
+			size_t wlen = 0;
+			while (wlen < size) {
+				size_t temp = stream->mBufferReader->read(buf + wlen, size - wlen);
+				if (temp == 0) {
+					// Can not read data, error occurred.
+					meddbg("StreamBufferReader::read failed!\n");
+					break;
+				}
+				wlen += temp;
+			}
+
+			if (worker->write(buf, size) < 0) {
+				// Error occurred, stop outputting
+				stream->mBufferWriter->setEndOfStream();
+				delete buf;
+				break;
+			}
+			delete buf;
+		}
+	}
+
+	return NULL;
+}
+
+void OutputHandler::sleepWorker()
+{
+	size_t spaces = mBufferWriter->sizeOfSpace();
+	std::unique_lock<std::mutex> lock(mMutex);
+	// In case of flushing or overrun, DO NOT sleep worker.
+	if (mIsWorkerAlive && !mIsFlushing && (spaces > 0)) {
+		mCondv.wait(lock);
+	}
+}
+
+void OutputHandler::wakenWorker()
+{
+	std::lock_guard<std::mutex> lock(mMutex);
+	mCondv.notify_one();
+}
+
+void OutputHandler::setStreamBuffer(std::shared_ptr<StreamBuffer> streamBuffer)
+{
+	if (mStreamBuffer) {
+		mStreamBuffer->setObserver(nullptr);
+		mBufferReader = nullptr;
+		mBufferWriter = nullptr;
+	}
+
+	mStreamBuffer = streamBuffer;
+
+	if (mStreamBuffer) {
+		mStreamBuffer->setObserver(this);
+		mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+		mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+	}
+}
+
+void OutputHandler::onBufferOverrun()
+{
+	auto mr = getRecorder();
+	if (mr) {
+		mr->notifyObserver(OBSERVER_COMMAND_BUFFER_OVERRUN);
+	}
+}
+
+void OutputHandler::onBufferUnderrun()
+{
+	auto mr = getRecorder();
+	if (mr) {
+		mr->notifyObserver(OBSERVER_COMMAND_BUFFER_UNDERRUN);
+	}
+}
+
+void OutputHandler::onBufferUpdated(ssize_t change, size_t current)
+{
+	medvdbg("OutputHandler::onBufferUpdated(%d, %u)\n", change, current);
+	if (change > 0) {
+		// Writing wake worker up
+		wakenWorker();
+	}
+}
+
+ssize_t OutputHandler::writeToStreamBuffer(unsigned char *buf, size_t size)
+{
+	if (buf == nullptr) {
+		meddbg("%s[line : %d] Fail : buf is nullptr\n");
```

Comment:
Did you omit "return (ssize_t)EOF;" here to let write() in line 327 return a negative value?

Reply:
Fixed it